### ISSUE TITLE
raco pkg install ignores the hash in the pkg source in favor of the hash stored in the catalog

### ISSUE THREAD

**Issue by sorawee** [racket/racket#4368]
**What version of Racket are you using?**
8.5 [cs]

When installing `unlike-assets-test`, the following message is shown:

```
$ raco pkg install unlike-assets-test
Resolving "unlike-assets-test" via https://pkgs.racket-lang.org
Downloading repository https://github.com/zyrolasting/unlike-assets.git?path=unlike-assets-test#7fb08d6902d34399c96a325d6d27f1ff84813b commit 7fb08d6902d34399c96a325d68fd27f1ff84813b
```

Note that 
```
7fb08d6902d34399c96a325d6d27f1ff84813b
``` 
and 
```
7fb08d6902d34399c96a325d68fd27f1ff84813b`
``` 
are not the same. The right one is correct, and has `8f` in the middle. The left one is wrong, and doesn't have `8f`.

`raco pkg install` uses `https://github.com/zyrolasting/unlike-assets.git?path=unlike-assets-test`. Ignores `#7fb08d69....` And uses `commit 7fb08d690....`. This is incorrect. It should instead use the hash in the package source `#7fb08d69....`.

**Comment by samth** [comment]
The reason the pkg server has a different checksum is that the current source fails to checkout (that hash does not exist in the repository). So the pkg server has kept the old checksum. 

However, I don't understand why this is working as intended. My expectation would be that it uses the checksum from the source, because that's the specification of the content, and then verifies that checksum against the checksum stored in the catalog (producing an error when they don't match). In other words, I don't see why the checksum is treated differently than any other part of the specification of the source, like the path component or the domain name.

**Comment by samth** [comment]
A different way to putting it: the following command succeeds, because the checksum is supplied, as in the case with the catalog on https://pkgs.racket-lang.org/.

```
raco pkg install --checksum 7fb08d6902d34399c96a325d68fd27f1ff84813b  https://github.com/zyrolasting/unlike-assets.git?path=unlike-assets-test#7fb08d6902d34399c96a325d6d27f1ff84813b 
```

I think that the `--checksum` argument, and checksums in general, should be projections (in the contract semantics sense). So nothing that fails without the checksum should be made to succeed, and certainly the checksum should not change what is actually installed.

**Comment by mflatt** [comment]
> I think that the --checksum argument, and checksums in general, should be projections (in the contract semantics sense). So nothing that fails without the checksum should be made to succeed, and certainly the checksum should not change what is actually installed.

I don't think this by itself fits with the way the pkg system and server currently work, where a Git-based package source might have something like `#main` with the intent that the "checksum" commit is updated automatically, and where a commit reported by the server is meant to be used even if the Git repo is updated meanwhile.

**Comment by sorawee** [comment]
- In what situation is "where a commit reported by the server is meant to be used even if the Git repo is updated meanwhile" useful? Personally I'd want that when I make a change, users can pull the change immediately.
- On the other hand, if we wish to preserve this behavior, `pkg-index` would need to set `checksum` to empty string when it cannot fetch a package source.

**Comment by samth** [comment]
To describe this in a little more detail after talking to @mflatt.

Right now, the semantics of a Git repository source (ignoring path specification) is something like this: 
- the source denotes a git repository as well as a specification for which commit in the repository to use (such as the one associated with a specific hash, branch, tag, or the default branch).
- a hash can also be provided (named "Checksum" but this is mislead)
- if the hash is separately provided, then it specifies the specific commit to be used for the content of the package
- if the hash is not provided at all, then the default commit from the original source specification is use. 

An alternative semantics would be the following:
- the source denotes a git repository as well as a commit
- a checksum can also be provided
- if the commit denoted by the source (possibly using various defaulting mechanisms) is not the same as the checksum (when provided) then `raco pkg install` errors.

The alternative semantics is more like what happens when using a source that denotes a ZIP file. Also, the alternative semantics is basically what a field called "Checksum" should mean.

However, the alternative semantics would break lots of things. In particular, right now if I push to the `master` branch of https://github.com/samth/fancy-app.git then the commit denoted by that package source changes, but the "Checksum" field at https://pkgs.racket-lang.org/package/fancy-app does not change immediately (that usually takes an hour or so). Under the alternative semantics, during that hour `raco pkg install fancy-app` would error because the checksum was different than what the package source denoted. That would be a bad state for the way Racket currently works.

It's worth noting that things are not as bad as they might seem. In particular, if a checksum is specified, you get a package whose content conforms to that checksum or an error. It's just that you can get that content in somewhat more situations than you might expect. 

It's possible that we could change things to make checksums actually checksums, but it would be a non-trivial change. One way to do it would be for pkgs.racket-lang.org to serve sources with hashes in them based on the checksum, even if that isn't the source the user specified. That would have lots of follow-on consequences, probably, and would require some deep thought before implementing. Probably other, larger changes would be possible as well.

**Comment by LiberalArtist** [comment]
I think one dimension of this question is that `raco pkg` and currently-existing public Git servers both happen to use SHA-1 in a way that lets the same 40-character hex string serve as a `raco pkg` "checsum" and as a Git "reference". Something is eventually going to change here, since [Git has a roadmap](https://git-scm.com/docs/hash-function-transition/) for transitioning to SHA-256 object ids.

Somewhat relatedly, there's some nuance to the fact that a Git reference can be either content-addressed and thus immutable (a commit object id) or symbolic and thus mutable (a branch or the default branch—or a tag, technically, though mutating a tag will make lots of people mad). There are some things I think are quite nice about the conventional way people tend to use Git package sources: in particular, tracking a branch (with a stable history) provides automatic updates while retaining a connection to history. (In contrast, for a ZIP file package source, you either needs to manually update a particular package catalog to look at a new url or change the content at the old url, making the old source inaccessible.)

I definitely think we want to keep good behavior for this scenario, where the Git reference in the package source is symbolic:

> However, the alternative semantics would break lots of things. In particular, right now if I push to the `master` branch of https://github.com/samth/fancy-app.git then the commit denoted by that package source changes, but the "Checksum" field at https://pkgs.racket-lang.org/package/fancy-app does not change immediately (that usually takes an hour or so). Under the alternative semantics, during that hour `raco pkg install fancy-app` would error because the checksum was different than what the package source denoted. That would be a bad state for the way Racket currently works.

In this specific situation, re <https://github.com/racket/racket/issues/4368#issuecomment-1190476482>:

> There's both a `Source` and a `Checksum` line above. So, `raco pkg install` is using the checksum that the pkg server reports, overriding the one in the source string. If you directly use `raco pkg install https://github.com/zyrolasting/unlike-assets.git?path=unlike-assets-test#7fb08d6902d34399c96a325d6d27f1ff84813b`, in contrast, you'll get the error (due to a wrong checksum) that you expect.
> 
> I think `raco pkg` is working as intended, and the issue is how the pkg server gets/reports a checksum.

Having a `Source` field that includes a Git object id (as opposed to a symbolic reference) and a `Checksum` field that conflicts with that object id seems like a bad state. I'd have to think more about the feasability and compatibility of distinguishing symbolic Git references from object ids (do we currently support abbreviated object ids?), but ideally I'd like to make that state an error, or at least make it difficult to get into that state.

If we were to consider broader changes, I think we'd need to figure out what purpose "checksums" are supposed to serve. If they're about transmission errors, to what extent do storage and transport layers (e.g. Git, HTTPS as opposed to HTTP) address those issues? If they're about authentication, are they the right tool for the job?

(For the later, a recent paper, ["Building a Secure Software Supply Chain with GNU Guix"](https://doi.org/10.22152/programming-journal.org/2023/7/1), presents an interesting authentication mechanism for Git commits in general.)

**Comment by samth** [comment]
> Having a Source field that includes a Git object id (as opposed to a symbolic reference) and a Checksum field that conflicts with that object id seems like a bad state. I'd have to think more about the feasability and compatibility of distinguishing symbolic Git references from object ids (do we currently support abbreviated object ids?), but ideally I'd like to make that state an error, or at least make it difficult to get into that state.

I agree with this. The way that the package under discussion here got into that state is as follows:

1. Set the source to something that works, which resolves to the checksum 7fb08d6902d34399c96a325d68fd27f1ff84813b
2. Update the source to something that doesn't work at all (such as the current Source field for the package) but which does work if the Git object is dropped.
3. Then, pkgs.racket-lang.org tries to update, fails, and keeps the old Checksum field but the new source field.

Hopefully this is a rare situation.

Of course, it would be possible to delete the Checksum on failure to resolve, or something else like that, but that seems fragile in the face of transient errors.
